hasprefix.go

```diff
@@ -0,0 +1,40 @@
+package vfs
+
+import (
+	"os"
+	"path/filepath"
+)
+
+// A Stater implements Stat.
+type Stater interface {
+	Stat(string) (os.FileInfo, error)
+}
+
+// HasPrefix returns true if p or any parent of p is the same file as prefix.
+// prefix must exist, but p may not. It is an expensive but accurate alternative
+// to the deprecated filepath.HasPrefix.
+func HasPrefix(fs Stater, p, prefix string) (bool, error) {
+	prefixFI, err := fs.Stat(prefix)
+	if err != nil {
+		return false, err
+	}
+	for {
+		fi, err := fs.Stat(p)
+		switch {
+		case err == nil:
+			if os.SameFile(fi, prefixFI) {
```

Comment:
Note that `os.SameFile` only works if the `os.FileInfo` was returned by the os-package. Meaning, your `Stater` interface doesn't fully capture the requirements of this function - if I implement it myself, with a custom `os.FileInfo`, this will probably fail.

Reply:
Good point, but I suggest leaving this unchanged. Reasoning:

* This package does not abstract `*os.File` (by design) and all operations fall back to the underlying `os` package. Therefore the `os.FileInfo` returned by `fs.Stat` should always be compatible with `os.SameFile`.

* Note that we have similar problems with methods like `os.IsNotExist` and `os.IsPermission`: these make assumptions about the `error` argument being in some way compatible with the underlying `os` package.

* At this point, I don't want to make `os.SameFile`, `os.IsNotExist`, and `os.IsPermission` methods of the `FS` interface to avoid cluttering it too much. But you're right, technically they should be.

Thanks again for the deep insight :)

Reply:
I added a comment about assumptions about `fs.Stat`.